File: 3.LinkedList/2.DLL/DLL.java
public class DLL {

    private Node head;
    private Node tail;
    private int size;

    public DLL() {
        this.size = 0;
    }

    // Insert at beginning
    public void insertFirst(int val) {
        Node node = new Node(val);
        node.next = head;
        node.prev = null;

        if (head != null) {
            head.prev = node;
        } else {
            tail = node;
        }

        head = node;
        size++;
    }
    public void insertLast(int val) {
    Node node = new Node(val);

    if (tail == null) { 
        head = node;
        tail = node;
        node.prev = null;
    } else {
        tail.next = node;
        node.prev = tail;
        tail = node;
    }

    tail.next = null;
    size++;
}
    public void insertslast(int val){
        Node node=new Node(val);
        Node last=head;
        if(head == null){
            node.prev=null;
            head=node;
            return;
        }
        while(last.next != null){
            last=last.next;
        }
        last.next=node;
        node.prev=last;
    }
    // Display forward
    public void insertAt(int val, int index) {
    if (index < 0 || index > size) {
        System.out.println("Index out of bounds");
        return;
    }

    if (index == 0) {
        insertFirst(val);
        return;
    }

    if (index == size) {
        insertLast(val);
        return;
    }

    // Insert in the middle
    Node temp = head;
    for (int i = 0; i < index - 1; i++) {
        temp = temp.next; // move to node before the insertion point
    }

    Node node = new Node(val, temp.next, temp); // new node points next to temp.next and prev to temp
    temp.next.prev = node; // update prev of the next node
    temp.next = node;       // link previous node to new node
    size++;
}

    public void display() {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.value + " <-> ");
            temp = temp.next;
        }
        System.out.println("End");
    }
    public void displayrev() {
        Node temp = tail;
        while (temp != null) {
            System.out.print(temp.value + " <-> ");
            temp = temp.prev;
        }
        System.out.println("End");
    }

    // Node class
    private class Node {
        private int value;
        private Node next;
        private Node prev;

        public Node(int value) {
            this.value = value;
        }

        public Node(int value, Node next, Node prev) {
            this.value = value;
            this.next = next;
            this.prev = prev;
        }
    }
}

File: 3.LinkedList/2.DLL/DLL.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None

class Main:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def insertFirst(self, val):
        node = Node(val)
        node.next = self.head
        node.prev = None

        if self.head is None:  
            self.tail = node
        else:
            self.head.prev = node

        self.head = node
        self.size += 1

    def display(self):
        temp = self.head
        while temp is not None:
            print(str(temp.value) + "<->", end="")
            temp = temp.next
        print("END")
    
    def displayrev(self):
        temp=self.tail
        while temp is not None:
            print(str(temp.value)+"<->",end=" ")
            temp=temp.prev
        print("Start")
    
    def deleteFirst(self):
        if self.head is None: 
            print("List is empty, nothing to delete.")
            return
        self.head = self.head.next
        if self.head is not None:
            self.head.prev = None
        else:
            self.tail = None

        self.size -= 1
    def deleteLast(self):
        if self.head is None:
            print("List is empty")
            return
        self.tail=self.tail.prev
        if self.head is not None:
            self.tail.next=None
        else:
            self.tail=None
        self.size-=1



my_list = Main()
my_list.insertFirst(50)
my_list.insertFirst(80)
my_list.insertFirst(90)
my_list.insertFirst(100)
my_list.display()
my_list.deleteLast()
my_list.display()
my_list.displayrev()

